### ISSUE TITLE
libdbd-sqlite3-perl FTBFS on mips: test failure

### ISSUE THREAD

**Issue by gregoa** [DBD-SQLite/DBD-SQLite#45]
We have the following bug reported to the Debian package of
DBD-SQLite (https://bugs.debian.org/919773):

It doesn't seem to be a bug in the packaging, so you may want to take
a look. Thanks!

Please note that there are some more details in the but report mentioned above.

------8<-----------8<-----------8<-----------8<-----------8<-----

```
Source: libdbd-sqlite3-perl
Version: 1.62-1
Severity: serious
Tags: ftbfs

https://buildd.debian.org/status/fetch.php?pkg=libdbd-sqlite3-perl&arch=mips&ver=1.62-1&stamp=1546291045&raw=0

...
t/65_db_config.t ...................................... 
1..79
ok 1 - An object of class 'DBI::db' isa 'DBI::db'
ok 2 - LOOKASIDE is not supported
ok 3
ok 4 - An object of class 'DBI::db' isa 'DBI::db'
ok 5 - LOOKASIDE is not supported
ok 6
ok 7 - An object of class 'DBI::db' isa 'DBI::db'
ok 8 - MAINDBNAME is not supported
ok 9
ok 10 - An object of class 'DBI::db' isa 'DBI::db'
ok 11 - MAINDBNAME is not supported
ok 12
ok 13 - An object of class 'DBI::db' isa 'DBI::db'
ok 14 - enable foreign key
ok 15 - disable foreign key
ok 16 - An object of class 'DBI::db' isa 'DBI::db'
ok 17 - enable foreign key
ok 18 - disable foreign key
ok 19 - An object of class 'DBI::db' isa 'DBI::db'
ok 20 - enable trigger
ok 21 - disable trigger
ok 22 - An object of class 'DBI::db' isa 'DBI::db'
ok 23 - enable trigger
ok 24 - disable trigger
ok 25 - An object of class 'DBI::db' isa 'DBI::db'
ok 26 - enable fts3_tokenizer
ok 27 - disable fts3_tokenizer
ok 28 - An object of class 'DBI::db' isa 'DBI::db'
ok 29 - enable fts3_tokenizer
ok 30 - disable fts3_tokenizer
ok 31 - An object of class 'DBI::db' isa 'DBI::db'
ok 32 - enable load_extension
ok 33 - disable load_extension
ok 34 - An object of class 'DBI::db' isa 'DBI::db'
ok 35 - enable load_extension
ok 36 - disable load_extension
ok 37 - An object of class 'DBI::db' isa 'DBI::db'
ok 38 - no checkpoint on close
ok 39 - checkpoint on close
ok 40 - An object of class 'DBI::db' isa 'DBI::db'
ok 41 - no checkpoint on close
ok 42 - checkpoint on close
ok 43 - An object of class 'DBI::db' isa 'DBI::db'
ok 44 - enable query planner stability guarantee
ok 45 - disable query planner stability guarantee
ok 46 - An object of class 'DBI::db' isa 'DBI::db'
ok 47 - enable query planner stability guarantee
ok 48 - disable query planner stability guarantee
ok 49 - An object of class 'DBI::db' isa 'DBI::db'
ok 50 - trigger explain query plan
ok 51 - no trigger explain query plan
ok 52 - An object of class 'DBI::db' isa 'DBI::db'
ok 53 - trigger explain query plan
ok 54 - no trigger explain query plan
ok 55 - An object of class 'DBI::db' isa 'DBI::db'
ok 56 - enable reset database
ok 57 - disable reset database
ok 58 - An object of class 'DBI::db' isa 'DBI::db'
ok 59 - enable reset database
ok 60 - disable reset database
ok 61 - An object of class 'DBI::db' isa 'DBI::db'
ok 62 - found sql
ok 63
ok 64 - updating sqlite_master is prohibited
ok 65
ok 66
ok 67 - updating sqlite_master is succeeded
ok 68 - sql not found
ok 69 - An object of class 'DBI::db' isa 'DBI::db'
ok 70 - found sql
ok 71
ok 72 - updating sqlite_master is prohibited
ok 73
ok 74
ok 75 - updating sqlite_master is succeeded
ok 76 - sql not found
Failed 3/79 subtests
...
Test Summary Report
-------------------
t/65_db_config.t                                    (Wstat: 11 Tests: 76 Failed: 0)
  Non-zero wait status: 11
  Parse errors: Bad plan.  You planned 79 tests but ran 76.
Files=111, Tests=3794, 61 wallclock secs ( 2.80 usr  0.48 sys + 48.42 cusr  4.10 csys = 55.80 CPU)
Result: FAIL
Failed 1/111 test programs. 0/3794 subtests failed.
make[1]: *** [Makefile:1086: test_dynamic] Error 255
```
------8<-----------8<-----------8<-----------8<-----------8<-----

Thanks for considering,
  gregor herrmann,
  Debian Perl Group

**Comment by charsbar** [comment]
Thanks but this is a duplicate of https://github.com/DBD-SQLite/DBD-SQLite/issues/44 (sorry I haven't released it yet)

**Comment by charsbar** [comment]
Sorry I overlooked "more details" (you mean the following, right?).

```
There is a difference earlier between mips and amd64: the first writes:

  # current DEFENSIVE value: 0

while the second has "1"

So it seems that these tests should be disabled for mips but it seems
that SQLITE_DBCONFIG_DEFENSIVE is badly set to a true value.
```

I don't have a mips environment so I can't tell for sure, but I hope https://github.com/DBD-SQLite/DBD-SQLite/commit/e5f82badfe26c007e5a4db6c9a9eb782936d917c would fix this.

**Comment by charsbar** [comment]
Hmm. Is there any test that also fails? And does https://github.com/DBD-SQLite/DBD-SQLite/commit/ce5b33b3812a445587a39e43d65f405b4a443e55 work? If only ```connect_ok(... sqlite_defensive => 1)``` dies/segfaults and other ```connect_ok()```s do not, then the only candidate should be https://github.com/DBD-SQLite/DBD-SQLite/blob/master/dbdimp.c#L462-L470 .

**Comment by gregoa** [comment]
All other tests pass.
If I add ce5b33b then `t/65_db_config.t`still fails in the same way.
I've asked the Debian mips porters for help, maybe they have some idea …

**Comment by charsbar** [comment]
Thank you. I'll try to look it again, but I don't have any idea for now.

**Comment by aurel32** [comment]
Given it fails exactly the same way on powerpc, I would say the issue is endianess related. It doesn't fail on 64-bit BE though, only 32-bit BE.

**Comment by gregoa** [comment]
@ntyni has come up with a patch which is available at https://bugs.debian.org/919773 and https://salsa.debian.org/perl-team/modules/packages/libdbd-sqlite3-perl/blob/master/debian/patches/0001-Fix-SQLITE_DBCONFIG_DEFENSIVE-parameter-types.patch
I'm going to upload the package now, and soon we'll see build results at https://buildd.debian.org/status/package.php?p=libdbd%2dsqlite3%2dperl (watch out for the 1.62-3 results).

**Comment by charsbar** [comment]
Thank you! Applied and shipped 1.63_03. @ntyni++
